sdk/src/main/scala/coop/rchain/sdk/dag/merging/DagMergingLogic.scala

```diff
@@ -0,0 +1,378 @@
+package coop.rchain.sdk.dag.merging
+
+import cats.Order
+import cats.syntax.all._
+import coop.rchain.sdk.dag.merging.data.Merge
+
+import scala.collection.compat.immutable.LazyList
+import scala.collection.immutable.Map
+import scala.collection.mutable
+import scala.math.Numeric.LongIsIntegral
+
+object DagMergingLogic {
+
+  /** All items in dependency chains. */
+  def withDependencies[D](of: Set[D], dependencyMap: Map[D, Set[D]]): Set[D] = {
+    def next(curOpt: Option[Set[D]]): Option[Set[D]] = curOpt.flatMap { c =>
+      val n = c.flatMap(dependencyMap.getOrElse(_, Set()))
+      n.nonEmpty.guard[Option].as(n)
+    }
+    LazyList.iterate(of.some)(next).takeWhile(_.nonEmpty).flatten.flatten.toSet
+  }
+
+  /** Deploys incompatible with finalized body. */
+  def incompatibleWithFinal[D](
+      acceptedFinally: Set[D],
+      rejectedFinally: Set[D],
+      conflictsMap: Map[D, Set[D]],
+      dependencyMap: Map[D, Set[D]]
+  ): Set[D] =
+    acceptedFinally.flatMap(conflictsMap) ++ rejectedFinally.flatMap(dependencyMap)
+
+  /** Split the scope into non overlapping partitions, greedily allocating intersecting chunks to bigger view. */
+  def partitionScope[D: Ordering](views: Seq[Set[D]]): Seq[Set[D]] = {
+    val r = LazyList.unfold(views) {
+      case (head @ seen) +: tail => (head, tail.map(_ -- seen)).some
+      case Seq()                 => none[(Set[D], List[Set[D]])]
+    }
+    r.toList
+  }
+
+  def computeRelationMap[D](
+      targetSet: Set[D],
+      sourceSet: Set[D],
+      relation: (D, D) => Boolean,
+      directed: Boolean
+  ): Map[D, Set[D]] =
+    targetSet.iterator
+      .flatMap(t => sourceSet.map((t, _)))
+      .foldLeft(mutable.Map.empty[D, Set[D]]) {
```

Comment:
Is there a reason to use mutable Map here?

Reply:
there are 2 conditional updates which look messy with updates of an immutable map. Also mutable should be a bit faster and the scope is quite small for use of mutable structure. So I found its handy to use here.